dynamodb/bucket.go

```diff
@@ -0,0 +1,158 @@
+	name                dbKey
+	capacity, remaining uint
+	reset               time.Time
+	rate                time.Duration
+	db                  db
+}
+
+// Capacity ...
+func (b *bucket) Capacity() uint {
+	return b.capacity
+}
+
+// Remaining space in the bucket.
+func (b *bucket) Remaining() uint {
+	return b.remaining
+}
+
+// Reset returns when the bucket will be drained.
+func (b *bucket) Reset() time.Time {
+	return b.reset
+}
+
+// Add to the bucket. This isn't a traditional leakybucket. It does not "drip" at a steady rate.
+// Rather, it treats the rate as a window for the specified capacity. If amount exceeds the
+// remaining capacity at any stage during the sliding window it will return `leakybucket.ErrorFull`
+// as expected
+func (b *bucket) Add(amount uint) (leakybucket.BucketState, error) {
+	// db.GetItem(bucket)
+	// if expired -> flush
+	// if (value + incrementAmount > capacity) -> error
+	// increment by amount -> return error if any
+	dbBucket, err := b.db.bucket(b.name)
+	if err != nil {
+		return b.state(), err
+	}
+	if dbBucket.isExpired() {
+		dbBucket, err = b.db.flushBucket(*dbBucket, time.Now().Add(b.rate))
+		if err != nil {
+			if err != errConcurrentFlush {
+				return b.state(), err
+			}
+			// if another consumer flushed the bucket re-fetch
+			dbBucket, err = b.db.bucket(b.name)
+			if err != nil {
+				return b.state(), err
+			}
+		}
+	}
+	// update local state
+	b.remaining = b.capacity - min(dbBucket.Value, b.capacity)
+	b.reset = dbBucket.Expiration
+	if amount > b.remaining {
+		return b.state(), leakybucket.ErrorFull
+	}
+	updatedDBBucket, err := b.db.incrementBucketValue(b.name, amount, b.capacity)
+	if err != nil {
+		if err == errBucketCapacityExceeded {
+			return b.state(), leakybucket.ErrorFull
+		}
+		return b.state(), err
+	}
+	// ensure we can't overflow
+	b.remaining = b.capacity - min(uint(updatedDBBucket.Value), b.capacity)
+	return b.state(), nil
+}
+
+func (b *bucket) state() leakybucket.BucketState {
+	return leakybucket.BucketState{
+		Capacity:  b.Capacity(),
+		Remaining: b.Remaining(),
+		Reset:     b.Reset(),
+	}
+}
+
+var _ leakybucket.Storage = &Storage{}
+
+// Storage is a dyanamodb-based, non thread-safe leaky bucket factory.
+type Storage struct {
+	db db
+}
+
+// Create a bucket. It will determine the current state of the bucket based on:
+// - The corresponding bucket in the database
+// - From scratch using the values provided
+func (s *Storage) Create(name string, capacity uint, rate time.Duration) (leakybucket.Bucket, error) {
+	bucket := &bucket{
+		name:      dbKey(name),
+		capacity:  capacity,
+		remaining: capacity,
+		reset:     time.Now().Add(rate),
+		rate:      rate,
+		db:        s.db,
+	}
+	dbBucket, err := s.db.bucket(bucket.name)
+	if err != nil {
+		// bubble up all errors other than a non-existing bucket
+		if err != errBucketNotFound {
+			return nil, err
+		}
+		if err := s.db.createBucket(newDDBBucket(name, bucket.reset)); err != nil {
+			return nil, err
```

Comment:
potential race here would lead to the user getting a conditional PutItem error here? I would expect this implementation to return the state of the bucket in the database if this is the case (e.g. `createBucket` might be better named `findOrCreateBucket` and handle the conditional put failure)

Reply:
https://github.com/Clever/leakybucket/pull/34/commits/acc900d900a63aa8d5b9933629f22b77d380ba99